modules/network/src/main/java/org/apache/ignite/internal/network/serialization/PerSessionSerializationService.java

```diff
@@ -182,6 +184,34 @@ public PerSessionSerializationService(@NotNull SerializationService serializatio
         return messages;
     }
 
+    private byte fieldFlags(FieldDescriptor fieldDescriptor) {
+        int bits = condMask(fieldDescriptor.isUnshared(), FieldDescriptorMessage.UNSHARED_MASK)
+                | condMask(fieldDescriptor.isPrimitive(), FieldDescriptorMessage.IS_PRIMITIVE)
+                | condMask(fieldDescriptor.isRuntimeTypeKnownUpfront(), FieldDescriptorMessage.IS_RUNTIME_TYPE_KNOWN_UPFRONT);
+        return (byte) bits;
+    }
+
+    private byte serializationAttributeFlags(Serialization serialization) {
+        int bits = condMask(serialization.hasWriteObject(), ClassDescriptorMessage.HAS_WRITE_OBJECT_MASK)
+                | condMask(serialization.hasReadObject(), ClassDescriptorMessage.HAS_READ_OBJECT_MASK)
+                | condMask(serialization.hasReadObjectNoData(), ClassDescriptorMessage.HAS_READ_OBJECT_NO_DATA_MASK)
+                | condMask(serialization.hasWriteReplace(), ClassDescriptorMessage.HAS_WRITE_REPLACE_MASK)
+                | condMask(serialization.hasReadResolve(), ClassDescriptorMessage.HAS_READ_RESOLVE_MASK);
+        return (byte) bits;
+    }
+
+    private int condMask(boolean value, int mask) {
```

Comment:
this method is always called multiple times for all the flags, maybe it would look better if it was something like:
```
condMask(boolean[] values, int[] masks)
```
It's arguably better, though

Reply:
Are you concerned with performance here? If yes, than allocation of 2 arrays per field message does not seem to be a good way to go.
If `condMask()` becomes hot, it will be inlined, so it would turn into the usual `int flags = (a & MASK1) + (a & MASK2) + ...` which is superfast. If `condMask()` is not hot, then we should not bother about it performance at all.

So I would leave it as it is now.